### DISCUSSION TITLE
How to log a Tensor Metric

### DISCUSSION THREAD

**Discussion by wellhowtosay** [Lightning-AI/pytorch-lightning#9007] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
I implement a Metric which returns a **Tensor** or a **Scalar** in compute, I try to log the Metric in *_step like:  
```
def test_step(self, batch, batch_idx):
        loss, pred, inp, target = self.compute_loss_and_acc(batch)
        metric = self.val_csc_metric(pred, inp, target)
        self.log("CSC_metric", self.val_csc_metric, on_epoch=True, on_step=True,logger=True)
```
but meet this:
pytorch_lightning.utilities.exceptions.MisconfigurationException: The metric `tensor([0.0541, 1.0000, 0.1026, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
        0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,    inf])` does not contain a single element, thus it cannot be converted to a scalar.
Exception ignored in: <function tqdm.__del__ at 0x7f8ba571f710>

I use self.log because I need it's on_epoch option to sync metric states among multi gpus at the end of eval/test epoch. 

How can I fix it or is there a alternative.
Thanks !

**Comment by talhaanwarch** [comment]
I think it would be `self.log("CSC_metric",metric , on_epoch=True, on_step=True,logger=True)` because val_csc_metric is a function and you want to log output of that function

**Comment by tchaton** [comment]
Yes, doing `metric = self.val_csc_metric(pred, inp, target)` performs a forward and metric is actually the result for the current batch.

However, you can provide the metric directly as suggested by @talhaanwarch.
